### DISCUSSION TITLE
credential.helper application algorithm

### DISCUSSION THREAD

**Discussion by andry81** [gitextensions/gitextensions#10731] [General] [answered]
I have found that the `credential.helper` executes backwards:

https://github.com/GitCredentialManager/git-credential-manager/issues/1116

```ini
credential.helper=wincred
credential.helper=
credential.helper=C:/Program\ Files\ \(x86\)/Git\ Credential\ Manager/git-credential-manager.exe
```
It makes confusion, why backwards?

**Comment by mstv** [comment, marked as answer]
`GitCredentialManager` is not part of Git Extensions itself. Please ask this question in the correct project.

**Reply by andry81** [reply]
The `credential.helper` reflects into `GitExtensions` not `GitCredentialManager` if change the order of values.

**Reply by RussKie** [reply]
The app invokes git and doesn't directly interact with GCM. If you can demonstrate that it's something the app is doing, then we'll be happy to look more into it.

**Reply by andry81** [reply]
If use this list:

```ini
credential.helper=wincred
credential.helper=
credential.helper=C:/Program\ Files\ \(x86\)/Git\ Credential\ Manager/git-credential-manager.exe
```

Then the order of application:
1. `GCM` and exit if succeeded.
2. `Windows Credential`

But if reverse the list:

```ini
credential.helper=C:/Program\ Files\ \(x86\)/Git\ Credential\ Manager/git-credential-manager.exe
credential.helper=
credential.helper=wincred
```

Then the order of application:
1. `Windows Credential` and exit if succeeded.
2. `GCM`

Which means is something in between of `GitExtensions` and the helpers does execute the helpers in reverse order.
Is that could be the `GitExtensions` itself?

**Reply by gerhardol** [reply]
GitExtensions does not use this setting, you get the same result when using Git from the command line.
It seems like Git adds the setting to a "stack".
